libs/core/executors/include/hpx/executors/parallel_scheduler.hpp

```diff
@@ -0,0 +1,285 @@
+        }
+
+        friend void tag_invoke(start_t, bulk_operation_state& os) noexcept
+        {
+            // std::cout << "Bulk start() called" << std::endl;
+            hpx::detail::try_catch_exception_ptr(
+                [&]() {
+                    thread_pool_scheduler exec{os.scheduler.get_thread_pool()};
+                    for (Shape i = 0; i < os.shape; ++i)
+                    {
+                        exec.execute([i, &os]() mutable {
+                            // std::cout << "Bulk task executing for index: " << i <<;
+                            os.f(i);
+                            if (--(*os.tasks_remaining) == 0)
+                            {
+                                // std::cout << "All bulk tasks completed" << std::endl;
+                                hpx::execution::experimental::set_value(
+                                    os.receiver);
+                            }
+                        });
+                    }
+                },
+                [&](std::exception_ptr ep) {
+                    // std::cerr << "Bulk error occurred" << std::endl;
+                    hpx::execution::experimental::set_error(
+                        HPX_MOVE(os.receiver), HPX_MOVE(ep));
+                });
+        }
+    };
+
+    template <typename Shape, typename F, typename Receiver>
+    auto tag_invoke(connect_t, bulk_sender<Shape, F>&& s, Receiver& r)
+    {
+        return bulk_operation_state<Receiver, Shape, F>{
+            HPX_MOVE(s.scheduler), r, s.shape, HPX_MOVE(s.f)};
+    }
+
+    struct parallel_sender
+    {
+        parallel_scheduler scheduler;
+
+#if defined(HPX_HAVE_STDEXEC)
+        using sender_concept = hpx::execution::experimental::sender_t;
+#endif
+        using completion_signatures =
+            hpx::execution::experimental::completion_signatures<
+                hpx::execution::experimental::set_value_t(),
+                hpx::execution::experimental::set_error_t(std::exception_ptr),
+                hpx::execution::experimental::set_stopped_t()>;
+
+        template <typename Env>
+        friend auto tag_invoke(
+            hpx::execution::experimental::get_completion_signatures_t,
+            parallel_sender const&, Env) noexcept -> completion_signatures
+        {
+            return {};
+        }
+
+        template <typename Receiver>
+        friend auto tag_invoke(connect_t, parallel_sender&& s, Receiver& r)
+        {
+            return operation_state<parallel_scheduler, Receiver>{
+                HPX_MOVE(s.scheduler), r};
+        }
+
+        template <typename Receiver>
+        friend auto tag_invoke(connect_t, parallel_sender& s, Receiver& r)
+        {
+            return operation_state<parallel_scheduler, Receiver>{
+                s.scheduler, r};
+        }
+
+        template <typename Shape, typename F>
+        friend auto tag_invoke(bulk_t, parallel_sender&& s, Shape shape, F&& f)
```

Comment:
For `bulk`, we have implemented an extension allowing to pass an arbitrary range as the `Shape` (in addition to the integral value), see here, for instance: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/execution/include/hpx/execution/algorithms/bulk.hpp#L273-L300. It might be a good idea to support this here as well.

Reply:
Also, you may want to move all types that are don't have to visible by users into `namespace detail` to avoid name clashes with other code.